File: DotNet/Messaging/MessageFetchers/MultiFetcher.cs
using System.Collections.Generic;
using System.Linq;
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Storage;

namespace Shastra.Hydra.Messaging.MessageFetchers
{
    public class MultiFetcher<TMessage> : IMessageFetcher<TMessage> where TMessage : TransportMessage
    {
        private readonly IMessageFetcher<TMessage>[] _fetchers;

        public MultiFetcher(params IMessageFetcher<TMessage>[] fetchers)
        {
            _fetchers = fetchers;
        }

        public IEnumerable<TMessage> MessagesAfterIdUpToSeq(IStore store, IMessageId startId, long lastSeq)
        {
            return _fetchers.AsParallel().Select(f => f.MessagesAfterIdUpToSeq(store, startId, lastSeq)).Merge();
        }

        public IEnumerable<TMessage> MessagesInSet(IStore store, IEnumerable<IMessageId> messageIds)
        {
            return _fetchers.AsParallel().Select(f => f.MessagesInSet(store, messageIds)).Merge();
        }
    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/messagefetchers/MultiFetcher.java
package uk.co.shastra.hydra.messaging.messagefetchers;

import java.util.ArrayList;

import uk.co.shastra.hydra.messaging.TransportMessage;
import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.storage.Store;
import uk.co.shastra.hydra.messaging.utils.ListUtils;

public class MultiFetcher<TMessage extends TransportMessage> implements MessageFetcher<TMessage> {

    private MessageFetcher<TMessage>[] fetchers;

	public MultiFetcher(MessageFetcher<TMessage>... fetchers)
    {
        this.fetchers = fetchers;
    }
    
	@Override
	public Iterable<TMessage> messagesAfterIdUpToSeq(Store store, MessageId startId, long lastSeq) {
		ArrayList<Iterable<TMessage>> messageLists = new ArrayList<Iterable<TMessage>>();
		for (MessageFetcher<TMessage> fetcher : fetchers) {
			messageLists.add(fetcher.messagesAfterIdUpToSeq(store, startId, lastSeq));
		}
		return ListUtils.Merge(messageLists);
	}

	@Override
	public Iterable<TMessage> messagesInSet(Store store, Iterable<MessageId> messageIds) {
		ArrayList<Iterable<TMessage>> messageLists = new ArrayList<Iterable<TMessage>>();
		for (MessageFetcher<TMessage> fetcher : fetchers) {
			messageLists.add(fetcher.messagesInSet(store, messageIds));
		}
		return ListUtils.Merge(messageLists);
	}

}
